pkg/autodiscovery/listeners/cloudfoundry.go

```diff
@@ -0,0 +1,244 @@
+	if err != nil {
+		return nil, err
+	}
+	return &CloudFoundryListener{
+		services:      map[string]Service{},
+		stop:          make(chan bool),
+		refreshTicker: time.NewTicker(10 * time.Second),
+		bbsCache:      bbsCache,
+	}, nil
+}
+
+// Listen periodically refreshes services from global BBS API cache
+func (l *CloudFoundryListener) Listen(newSvc chan<- Service, delSvc chan<- Service) {
+	// setup the I/O channels
+	l.newService = newSvc
+	l.delService = delSvc
+
+	go func() {
+		l.refreshServices(true)
+		for {
+			select {
+			case <-l.stop:
+				l.refreshTicker.Stop()
+				return
+			case <-l.refreshTicker.C:
+				l.refreshServices(false)
+			}
+		}
+	}()
+}
+
+func (l *CloudFoundryListener) refreshServices(firstRun bool) {
+	log.Debug("Refreshing services via CloudFoundryListener")
+	allActualLRPs, desiredLRPs := l.bbsCache.GetAllLRPs()
+
+	// if not found and running, add it
+	// at the end, compare what we saw and what is cached and kill what's not there anymore
+	notSeen := make(map[string]interface{})
+	for i := range l.services {
+		notSeen[i] = nil
+	}
+
+	adIdentifiers := l.getAllADIdentifiers(desiredLRPs, allActualLRPs)
+	for _, id := range adIdentifiers {
+		strID := id.String()
+		if _, found := l.services[strID]; found {
+			// delete is no-op when we try to delete a key that doesn't exist
+			// NOTE: this will remove old versions of services on redeploys because ADIdentifier contains ProcessGUID,
+			//       which changes by redeploying
+			delete(notSeen, strID)
+			continue
+		}
+		svc := l.createService(id, firstRun)
+		// if the container is not in RUNNING state, we can't populate ports, thus createService would return nil
+		if svc != nil {
+			l.newService <- svc
+		}
+	}
+
+	for adID := range notSeen {
+		l.Lock()
+		l.delService <- l.services[adID]
+		delete(l.services, adID)
+		l.Unlock()
+	}
+}
+
+func (l *CloudFoundryListener) createService(adID cloudfoundry.ADIdentifier, firstRun bool) *CloudFoundryService {
+	l.Lock()
+	defer l.Unlock()
+	crTime := integration.After
+	if firstRun {
+		crTime = integration.Before
+	}
+
+	var svc *CloudFoundryService
+	aLRP := adID.GetActualLRP()
+	if aLRP == nil {
+		// non-container service
+		// NOTE: non-container services intentionally have no IPs or ports, everything is supposed to be configured
+		// through the "variables" section in the AD configuration
+		svc = &CloudFoundryService{
+			adIdentifier:   adID,
+			containerIPs:   map[string]string{},
+			containerPorts: []ContainerPort{},
+			creationTime:   crTime,
+		}
+	} else {
+		if aLRP.State != cloudfoundry.ActualLrpStateRunning {
+			return nil
+		}
+		// container service => we need one service per container instance
+		ips := map[string]string{CfServiceContainerIP: aLRP.ContainerIP}
+		ports := []ContainerPort{}
+		for _, p := range aLRP.Ports {
+			ports = append(ports, ContainerPort{
+				// TODO: because of how configresolver.getPort works, we can't use e.g. port_8080, so we use port_p8080
```

Comment:
remove

Reply:
I'll change it to a NOTE, since it's still noteworthy :)